opnode/rollup/driver/driver.go

```diff
@@ -2,71 +2,136 @@ package driver
-	// An error is returned if the sync starting point could not be determined (due to timeouts, wrong-chain, etc.)
-	findSyncStart(ctx context.Context) (nextRefL1 eth.BlockID, refL2 eth.BlockID, err error)
-	// driverStep explicitly calls the engine to derive a L1 block into a L2 block, and apply it on top of the given L2 block.
-	// The finalized L2 block is provided to update the engine with finality, but does not affect the derivation step itself.
-	// The resulting L2 block ID is returned, or an error if the derivation fails.
-	driverStep(ctx context.Context, nextRefL1 eth.BlockID, refL2 eth.BlockID, finalized eth.BlockID) (l2ID eth.BlockID, err error)
-}
+// keep making many sync steps if we can make sync progress
+const hot = time.Millisecond * 30
 
-type EngineDriver struct {
-	Log log.Logger
-	// API bindings to execution engine
-	RPC     DriverAPI
-	DL      Downloader
-	SyncRef rollupSync.SyncReference
+// check on sync regularly, but prioritize sync triggers with head updates etc.
+const cold = time.Second * 8
 
-	// The current driving force, to shutdown before closing the engine.
-	driveSub ethereum.Subscription
-	// There may only be 1 driver at a time
-	driveLock sync.Mutex
+// at least try every minute to sync, even if things are going well
+const max = time.Minute
 
-	EngineDriverState
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
 }
 
-func (e *EngineDriver) Drive(ctx context.Context, l1Heads <-chan eth.HeadSignal) ethereum.Subscription {
-	e.driveLock.Lock()
-	defer e.driveLock.Unlock()
-	if e.driveSub != nil {
-		return e.driveSub
+type Driver struct {
+	log     log.Logger
+	rpc     DriverAPI
+	syncRef rollupSync.SyncReference
+	dl      Downloader
+	l1Heads <-chan eth.HeadSignal
+	done    chan chan error
+	genesis rollup.Genesis
+	state   // embedded engine state
+}
+
+func NewDriver(l2 DriverAPI, l1 l1.Source, log log.Logger, genesis rollup.Genesis) *Driver {
+	return &Driver{
+		log:     log,
+		rpc:     l2,
+		syncRef: rollupSync.SyncSource{L1: l1, L2: l2},
+		dl:      l1,
+		done:    make(chan chan error),
+		genesis: genesis,
+		state:   state{Genesis: genesis},
 	}
+}
 
-	e.driveSub = event.NewSubscription(NewDriverLoop(ctx, &e.EngineDriverState, e.Log, l1Heads, e))
-	return e.driveSub
+func (d *Driver) Start(ctx context.Context, l1Heads <-chan eth.HeadSignal) error {
+	d.l1Heads = l1Heads
+	if !d.requestUpdate(ctx, d.log, d) {
+		d.log.Error("failed to fetch engine head, defaulting to genesis")
+		d.updateHead(d.Genesis.L1, d.Genesis.L2)
+	}
+	go d.loop()
+	return nil
+}
+func (d *Driver) Close() error {
+	ec := make(chan error)
+	d.done <- ec
+	err := <-ec
+	return err
 }
 
-func (e *EngineDriver) requestEngineHead(ctx context.Context) (refL1 eth.BlockID, refL2 eth.BlockID, err error) {
-	refL1, refL2, _, err = e.SyncRef.RefByL2Num(ctx, nil, &e.Genesis)
-	return
+func (d *Driver) loop() {
```

Comment:
Tbh I liked how the loop was separated because it makes encapsulated testing a bit easier. It had a lot of similar function arguments though, so I'm ok with attaching it to the Driver.

Reply:
It was enough overlap to where the driver doesn't have much that the loop doesn't depend on, but if the driver keeps growing, we can split the loop out later.